File: 3093-longest-common-suffix-queries/solution.cpp
#include <vector>
#include <string>
#include <cstring>

using namespace std;

// Flat, pool-allocated Trie Node structure for high performance
struct Node {
    int children[26];
    int best_index;
};

class Solution {
private:
    vector<Node> trie;

    int create_node() {
        Node n;
        memset(n.children, -1, sizeof(n.children));
        n.best_index = -1;
        trie.push_back(n);
        return trie.size() - 1;
    }

    void insert(const string& s, int idx, const vector<string>& wordsContainer) {
        int curr = 0; // Root is always at index 0
        
        // Helper lambda to update the best index at a node
        auto update_best = [&](int node_idx) {
            int& best = trie[node_idx].best_index;
            if (best == -1 || wordsContainer[idx].length() < wordsContainer[best].length()) {
                best = idx;
            }
        };

        update_best(curr);
        
        // Traverse the string in reverse order to build the suffix trie
        for (int i = s.length() - 1; i >= 0; --i) {
            int c = s[i] - 'a';
            if (trie[curr].children[c] == -1) {
                trie[curr].children[c] = create_node();
            }
            curr = trie[curr].children[c];
            update_best(curr);
        }
    }

    int query(const string& s) {
        int curr = 0;
        int best = trie[curr].best_index;
        
        // Match characters in reverse order
        for (int i = s.length() - 1; i >= 0; --i) {
            int c = s[i] - 'a';
            if (trie[curr].children[c] == -1) {
                break;
            }
            curr = trie[curr].children[c];
            best = trie[curr].best_index;
        }
        return best;
    }

public:
    vector<int> stringIndices(vector<string>& wordsContainer, vector<string>& wordsQuery) {
        trie.clear();
        create_node(); // Create root node
        
        for (int i = 0; i < wordsContainer.size(); ++i) {
            insert(wordsContainer[i], i, wordsContainer);
        }
        
        vector<int> ans;
        ans.reserve(wordsQuery.size());
        for (const string& q : wordsQuery) {
            ans.push_back(query(q));
        }
        
        return ans;
    }
};

File: 3093-longest-common-suffix-queries/solution.py
from typing import List

class TrieNode:
    def __init__(self):
        self.children = {}
        self.best_index = -1

class Solution:
    def stringIndices(self, wordsContainer: List[str], wordsQuery: List[str]) -> List[int]:
        root = TrieNode()
        
        def update_best(node: TrieNode, idx: int) -> None:
            if node.best_index == -1:
                node.best_index = idx
            elif len(wordsContainer[idx]) < len(wordsContainer[node.best_index]):
                node.best_index = idx
                
        # Build the suffix trie
        for idx, word in enumerate(wordsContainer):
            curr = root
            update_best(curr, idx)
            # Traverse characters in reverse
            for char in reversed(word):
                if char not in curr.children:
                    curr.children[char] = TrieNode()
                curr = curr.children[char]
                update_best(curr, idx)
                
        ans = []
        # Answer each query
        for query in wordsQuery:
            curr = root
            best = root.best_index
            for char in reversed(query):
                if char not in curr.children:
                    break
                curr = curr.children[char]
                best = curr.best_index
            ans.append(best)
            
        return ans